pkg/receive/handler.go

```diff
@@ -533,28 +533,39 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	}
 	h.mtx.RUnlock()
 
-	var err error
-	ctx, cancel := context.WithTimeout(ctx, h.options.ForwardTimeout)
-	defer func() {
-		// If there is no error, let forward requests optimistically run until timeout.
+	var (
+		quorum       = h.writeQuorum()
+		fctx, cancel = context.WithTimeout(context.Background(), h.options.ForwardTimeout) //nolint:govet
+		errChan      = make(chan error)
+	)
+	go func() {
+		defer close(errChan)
+		errChan <- h.fanoutForward(fctx, tenant, replicas, wreqs, quorum)
+	}()
+
+	select {
+	case err := <-errChan:
 		if err != nil {
+			// fanoutForward only returns an error if successThreshold (quorum) is not reached.
+			// Cancel all in-flight requests.
 			cancel()
+
+			if countCause(err, isNotReady) >= quorum {
+				return errors.Wrap(tsdb.ErrNotReady, "replicate: quorum not reached")
+			}
+			if countCause(err, isConflict) >= quorum {
+				return errors.Wrap(conflictErr, "replicate: quorum not reached")
+			}
+
+			return errors.Wrap(err, "unexpected error")
 		}
-	}()
 
-	quorum := h.writeQuorum()
-	err = h.fanoutForward(ctx, tenant, replicas, wreqs, quorum)
-	if countCause(err, isNotReady) >= quorum {
-		return tsdb.ErrNotReady
-	}
-	if countCause(err, isConflict) >= quorum {
-		return errors.Wrap(conflictErr, "did not meet success threshold due to conflict")
+		// If there is no error, quorum reached, let forward requests optimistically run until timeout.
+		return nil //nolint:govet
```

Comment:
I don't think I understand why the `nolint` comment is necessary here? The `err != nil` branch is already covered, and will return no matter what the `err` value is. I feel like I'm missing something.

Reply:
`govet` here complains that for this return branch the context is not cancelled.

Reply:
I see, can you explain exactly this in the comment? The connection between the comment above it and the nolint wasn't obvious to me (maybe it should have been but I would guess that future me would make the same mistake)

Reply:
Added comments to explain the reasoning behind `nolint`s.

Reply:
I apologize for being picky on this, but for the sake of it being extremely obvious, can you add to the comments something along the lines of "which is why the no-lint comment is required and intentional here."